File: dcr/tokens.py
"""Token accounting.

`B_attention` is only meaningful if costs are measured in the same unit the
model bills in. The default estimator is a cheap character heuristic so the
runtime works offline; swap in a real tokenizer (or the Anthropic
`messages.count_tokens` endpoint) by passing any callable of `str -> int`.
"""

from __future__ import annotations

from typing import Callable

TokenEstimator = Callable[[str], int]

CHARS_PER_TOKEN = 4


def estimate_tokens(text: str) -> int:
    if not text:
        return 0
    return max(1, -(-len(text) // CHARS_PER_TOKEN))


def make_count_tokens_estimator(client, model: str = "claude-opus-5") -> TokenEstimator:
    """Exact estimator backed by the Anthropic token-counting endpoint.

    Only worth using when budgets are tight enough that heuristic error
    matters; it costs a network round trip per call, so cache aggressively.
    """

    cache: dict[str, int] = {}

    def count(text: str) -> int:
        if text in cache:
            return cache[text]
        result = client.messages.count_tokens(
            model=model, messages=[{"role": "user", "content": text or " "}]
        )
        cache[text] = result.input_tokens
        return result.input_tokens

    return count

File: src/tokens.rs
//! Token accounting.
//!
//! `B_attention` is only meaningful if costs are measured in the same unit the
//! model bills in. The default estimator is a cheap character heuristic so the
//! runtime works offline; swap in a real tokenizer by setting
//! [`Estimator::Custom`].

pub const CHARS_PER_TOKEN: usize = 4;

pub fn estimate_tokens(text: &str) -> usize {
    if text.is_empty() {
        return 0;
    }
    text.len().div_ceil(CHARS_PER_TOKEN).max(1)
}

/// How the runtime prices text.
#[derive(Default)]
pub enum Estimator {
    /// `len / 4`, no network, good enough to keep a budget honest.
    #[default]
    Heuristic,
    /// Anything exact — a real tokenizer, or the Anthropic token-counting
    /// endpoint behind a cache.
    Custom(Box<dyn Fn(&str) -> usize>),
}

impl Estimator {
    pub fn count(&self, text: &str) -> usize {
        match self {
            Estimator::Heuristic => estimate_tokens(text),
            Estimator::Custom(f) => f(text),
        }
    }
}

impl std::fmt::Debug for Estimator {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Estimator::Heuristic => f.write_str("Estimator::Heuristic"),
            Estimator::Custom(_) => f.write_str("Estimator::Custom(..)"),
        }
    }
}
